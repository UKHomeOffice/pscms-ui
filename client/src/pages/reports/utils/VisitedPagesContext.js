import React, { createContext, useState } from 'react';
import PropTypes from 'prop-types';

export const VisitedPagesContext = createContext({
  visitedPages: {},
  setVisitedPages: () => {},
});

export const VisitedPagesContextProvider = ({ children }) => {
  const [visitedPages, setVisitedPages] = useState({});

  return (
    <VisitedPagesContext.Provider value={{ visitedPages, setVisitedPages }}>
      {children}
    </VisitedPagesContext.Provider>
  );
};

VisitedPagesContextProvider.propTypes = {
  children: PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.node), PropTypes.node]).isRequired,
};
