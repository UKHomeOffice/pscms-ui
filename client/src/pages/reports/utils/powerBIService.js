import { factories, service } from 'powerbi-client';

const powerbi = new service.Service(
  factories.hpmFactory,
  factories.wpmpFactory,
  factories.routerFactory
);

export default powerbi;
