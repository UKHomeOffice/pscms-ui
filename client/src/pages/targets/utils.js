import dayjs from 'dayjs';

export const formatAddress = (address) => {
  if (!address) {
    return null;
  }
  const formattedAddress =
    (address.line1 ? `${address.line1}\n` : '') +
    (address.line2 ? `${address.line2}\n` : '') +
    (address.line3 ? `${address.line3}\n` : '') +
    (address.city ? `${address.city}\n` : '') +
    (address.postCode ? `${address.postCode}\n` : '') +
    (address.country ? `${address.country}\n` : '');
  return formattedAddress;
};

export const formatCategory = (category) => {
  return category.slice(-1);
};

export const formatDate = (date) => {
  return dayjs(date).format('DD MM YYYY HH:mm');
};

export const formatPerson = (firstName, middleName, lastName, docNumber) => {
  const name =
    (firstName ? `${firstName} ` : '') +
    (middleName ? `${middleName} ` : '') +
    (lastName ? `${lastName}` : '') +
    (docNumber ? `, ${docNumber}` : '');
  return name;
};
