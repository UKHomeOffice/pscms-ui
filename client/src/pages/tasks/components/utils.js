import dayjs from 'dayjs';
import relativeTime from 'dayjs/plugin/relativeTime';

dayjs.extend(relativeTime);

export const cleanSubmissionData = (submissionData) => {
  const propsToKeep = [
    'name',
    'description',
    'priority',
    'assignee',
    'owner',
    'delegationState',
    'due',
    'followUp',
    'parentTaskId',
    'caseInstanceId',
    'tenantId',
  ];
  const result = {};

  propsToKeep.forEach((key) => {
    result[key] = submissionData[key];
  });
  return result;
};

export const isOverDue = (due) => {
  if (dayjs(due).fromNow().includes('ago')) {
    return (
      <span
        aria-label={`due ${dayjs(due).fromNow()}`}
        className="govuk-!-font-size-19 govuk-!-font-weight-bold overdue"
      >
        {`Overdue ${dayjs(due).fromNow()}`}
      </span>
    );
  }
  return (
    <span
      aria-label={`due ${dayjs(due).fromNow()}`}
      className="govuk-!-font-size-19 govuk-!-font-weight-bold not-overdue"
    >
      {`Due ${dayjs(due).fromNow()}`}
    </span>
  );
};

export const isDateValid = (date, format) => dayjs(date).format(format) === date;
