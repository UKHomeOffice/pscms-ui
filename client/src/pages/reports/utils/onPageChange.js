import { powerBiSchema as $schema } from '../../../utils/constants';

const onPageChange = (branchName, event, visitedPages) => {
  let target;
  let visualNumber;
  const {
    newPage,
    newPage: { displayName },
  } = event.detail;

  if (visitedPages.current.includes(displayName)) return;

  visitedPages.current.push(displayName);

  if (displayName === 'Command Brief - OAR') {
    visualNumber = 5;
    target = { table: 'OAR', column: 'Branch Name' };
  } else if (displayName === 'Command Brief - IEN') {
    visualNumber = 4;
    target = { table: 'IEN', column: 'Branch' };
  } else {
    return;
  }

  newPage.getVisuals().then((visuals) => {
    const targetVisual = visuals[visualNumber];
    targetVisual
      .setSlicerState({
        filters: [
          {
            $schema,
            target,
            operator: 'In',
            values: [branchName],
          },
        ],
      })
      .catch((errors) => {
        // eslint-disable-next-line no-console
        console.log('Errors loading visuals:', errors);
      });
  });
};

export default onPageChange;
