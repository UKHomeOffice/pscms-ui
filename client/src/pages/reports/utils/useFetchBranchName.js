import { useEffect } from 'react';
import axios from 'axios';
import { useAxios, useIsMounted } from '../../../utils/hooks';

const useFetchBranchName = async ({ branchId, setBranchName }) => {
  const axiosInstance = useAxios();
  const isMounted = useIsMounted();
  useEffect(() => {
    const source = axios.CancelToken.source();
    if (axiosInstance && branchId) {
      const fetchData = async () => {
        try {
          const response = await axiosInstance.get(
            `/refdata/v2/entities/branch?filter=id=eq.${branchId}`,
            {
              cancelToken: source.token,
            }
          );
          if (isMounted.current) {
            const branchName = response.data.data.length && response.data.data[0].name;
            setBranchName(branchName);
          }
        } catch (e) {
          // eslint-disable-next-line no-console
          console.log('Error fetching branch name:', e);
        }
      };
      fetchData();
    }
    return () => {
      source.cancel('Cancelling request');
    };
  }, [axiosInstance, branchId, isMounted, setBranchName]);
};

export default useFetchBranchName;
