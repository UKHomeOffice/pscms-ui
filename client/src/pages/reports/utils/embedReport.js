import { models } from 'powerbi-client';
import { powerBIBranchNames } from '../../../utils/constants';
import onPageChange from './onPageChange';
import powerbi from './powerBIService';

const embedReport = ({
  accessToken,
  branchName,
  embedUrl,
  id,
  mobileLayout,
  reportContainer,
  visitedPages,
}) => {
  let report;
  if (reportContainer) {
    const embedConfig = {
      accessToken,
      type: 'report',
      embedUrl,
      id,
      permissions: models.Permissions.Read,
      settings: {
        filterPaneEnabled: false,
        layoutType: mobileLayout ? models.LayoutType.MobilePortrait : models.LayoutType.Master,
      },
      tokenType: models.TokenType.Embed,
    };

    powerbi.reset(reportContainer.current);
    report = powerbi.embed(reportContainer.current, embedConfig);
    if (powerBIBranchNames.includes(branchName)) {
      report.on('pageChanged', (event) => onPageChange(branchName, event, visitedPages));
    }
  }
  return report;
};

export default embedReport;
