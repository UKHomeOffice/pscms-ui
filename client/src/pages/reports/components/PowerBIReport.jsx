import React, { useContext, useEffect, useRef, useState } from 'react';
import { useNavigation } from 'react-navi';
import styled from 'styled-components';
import { useMatomo } from '@datapunt/matomo-tracker-react';
import LogoBar from './LogoBar';
import { TeamContext } from '../../../utils/TeamContext';
import { mobileWidth } from '../../../utils/constants';
import useFetchBranchName from '../utils/useFetchBranchName';
import embedReport from '../utils/embedReport';

const PowerBIReport = () => {
  const navigation = useNavigation();
  const state = navigation.extractState();
  const reportContainer = useRef(null);
  const [branchName, setBranchName] = useState(null);
  const [embedReportArgs, setEmbedReportArgs] = useState(null);
  const [mobileLayout, setMobileLayout] = useState(null);
  const [report, setReport] = useState(null);
  const branchId = useContext(TeamContext).team?.branchid;
  const visitedPages = useRef([]);
  const { trackPageView } = useMatomo();

  useEffect(() => {
    trackPageView();
  }, []);

  useEffect(() => {
    if (!state) {
      navigation.navigate('/reports');
    } else {
      setMobileLayout(window.innerWidth < mobileWidth);
      setEmbedReportArgs({
        mobileLayout,
        reportContainer,
        visitedPages,
        ...state,
      });
    }
  }, [mobileLayout, navigation, reportContainer, state, visitedPages]);

  useFetchBranchName({
    branchId,
    setBranchName,
  });

  useEffect(() => {
    setReport(embedReport({ branchName, ...embedReportArgs }));
  }, [branchName, embedReportArgs]);

  return (
    <ReportContainer
      {...{ mobileLayout }}
      style={{
        height: mobileLayout ? '50vh' : '50vw',
        maxHeight: mobileLayout ? null : '70vmin',
      }}
    >
      <Report id="report" ref={reportContainer} />
      {report && !mobileLayout ? <LogoBar {...{ report, branchName, visitedPages }} /> : null}
    </ReportContainer>
  );
};

export const ReportContainer = styled.div`
  height: ${(props) => (props.mobileLayout ? '50vh' : '50vw')};
  maxheight: ${(props) => (props.mobileLayout ? null : '70vmin')};
  position: relative;
`;

const Report = styled.div`
  width: 100%;
  height: 100%;
`;

export default PowerBIReport;
