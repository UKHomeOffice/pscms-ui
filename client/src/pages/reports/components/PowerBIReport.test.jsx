import React from 'react';
import PropTypes from 'prop-types';
import { mount } from 'enzyme';
import 'jest-styled-components';
import { useNavigation } from 'react-navi';
import PowerBIReport, { ReportContainer } from './PowerBIReport';
import { mockNavigate } from '../../../setupTests';
import { TeamContext } from '../../../utils/TeamContext';

describe('PowerBIReport Page', () => {
  const extractState = () => ({
    accessToken: 'xxx',
    embedUrl: 'http://www.example.com',
    id: 'abc',
    name: 'Power BI Report',
  });

  it('renders without crashing', () => {
    useNavigation.mockImplementationOnce(() => ({
      extractState,
    }));
    const wrapper = mount(<PowerBIReport />);
    expect(wrapper.exists()).toBe(true);
  });

  it('redirects if no state found', () => {
    mount(<PowerBIReport />);
    expect(mockNavigate).toHaveBeenCalled();
  });

  it('renders report element when branchid is found', async () => {
    useNavigation.mockImplementationOnce(() => ({
      extractState,
    }));

    const { Provider } = TeamContext;
    const TestComponent = ({ children }) => (
      <Provider value={{ team: { branchid: 23 } }}>{children}</Provider>
    );

    TestComponent.propTypes = {
      children: PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.node), PropTypes.node]).isRequired,
    };

    const wrapper = mount(
      <TestComponent>
        <PowerBIReport />
      </TestComponent>
    );
    expect(wrapper.find('#report').exists()).toBe(true);
  });

  it('renders styles as expected', () => {
    const wrapper = mount(<ReportContainer mobileLayout />);
    expect(wrapper).toHaveStyleRule('height', '50vh');
  });

  it('matches snapshot', () => {
    const wrapper = mount(<PowerBIReport />);
    expect(wrapper).toMatchSnapshot();
  });
});
