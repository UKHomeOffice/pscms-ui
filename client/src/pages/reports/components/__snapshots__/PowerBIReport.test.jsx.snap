// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PowerBIReport Page matches snapshot 1`] = `
.c0 {
  height: 50vw;
  maxheight: 70vmin;
  position: relative;
}

.c1 {
  width: 100%;
  height: 100%;
}

<PowerBIReport>
  <styled.div
    mobileLayout={null}
    style={
      Object {
        "height": "50vw",
        "maxHeight": "70vmin",
      }
    }
  >
    <div
      className="c0"
      style={
        Object {
          "height": "50vw",
          "maxHeight": "70vmin",
        }
      }
    >
      <styled.div
        id="report"
      >
        <div
          className="c1"
          id="report"
        />
      </styled.div>
    </div>
  </styled.div>
</PowerBIReport>
`;
